app/controllers/budgets_controller.rb

```diff
@@ -52,6 +53,11 @@ def initialize_budgets_if_needed
     @grant.save if @grant.changed?
   end
 
+  def set_grant
+    @grant = Grant.visible_for_user(current_user).where(id: params[:grant_id]).first
+    not_found_error if @grant.nil?
+  end
+
```

Comment:
This seems like controller logic to me.  Find a grant that's visible with an id of X.  If not there, take them to the 404 page.  If `visible_for_user` was implemented here... that'd be a different story.